File: 백준/Silver/2512. 예산/예산.java
import java.util.Arrays;
import java.util.Scanner;

class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int[] arr = new int[N];
		for(int i = 0; i<N; i++) {
			arr[i] = sc.nextInt();
		}
		// 
		int max = sc.nextInt(); 
		int ans = 0;
		
		// 이분탐색이에욥 정력 해줘요
		Arrays.sort(arr);
		
		int end = arr[N-1];
		int start = 0;
		while(start<=end) {
			int mid = (start+end)/2;
			int sum = 0;
			for(int i = 0; i<arr.length; i++) {
				if(arr[i] >= mid) sum+= mid; // 주어진 예산의 값이 배정된 예산의 값보다 크면 배정된 예산의 값으로 갱신한다.
				else sum += arr[i]; // 배정된 예산 값이 더 작으면 주어진 예산을 그냥 더해
			}
			if(sum>max) end = mid-1; // 총 예산보다 sum값이 클 수 없다. 상한 수정 필요
			else {
				start = mid +1; 
				ans = Math.max(ans, mid); //정해진 총액 이하에서 가능한 한 최대의 총 예산을 배정해야한다.
			}
		}System.out.println(ans);
	}
}

File: 백준/Silver/2512. 예산/예산.py
# 1. 예산 요청 중 최댓값을 상한선, 1을 하한선으로 상한액 찾기
# 2. 예산 요청과 상한액 비교해서 상한액보다 작거나 같으면 요청 금액 그대로
# 3. 상한액보다 크면 상한액 배정
# 4. 배정된 예산의 합을 전체 예산과 비교

import sys
input = sys.stdin.readline

n = int(input())
array = list(map(int, input().split()))
m = int(input())

# 최댓값이 상한선
# 1이 하한선
start, end = 1, max(array)
while start<= end:
    # 배정 예산의 총액이 전체 예산보다 크면 상한액 낮추기
    # 배정 예산의 총액이 전체 예산보다 작거나 같으면 상한액 높이기
    mid = (start+end)//2
    sum_budget = 0
    for r in array:
        sum_budget += min(mid, r)
    if sum_budget > m:
        end = mid - 1
    elif sum_budget <= m:
        start = mid +  1

print(end)